sycl/source/handler.cpp

```diff
@@ -108,8 +108,13 @@ event handler::finalize() {
   if (getCGTypeVersion(MCGType) >
       static_cast<unsigned int>(detail::CG::CG_VERSION::V0)) {
     // If there were uses of set_specialization_constant build the kernel_bundle
+    bool HasSpecConstArgs =
+        std::any_of(MArgs.begin(), MArgs.end(), [](detail::ArgDesc Arg) {
+          return Arg.MType == detail::kernel_param_kind_t::
+                                  kind_specialization_constants_buffer;
+        });
     std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr =
-        getOrInsertHandlerKernelBundle(/*Insert=*/false);
+        getOrInsertHandlerKernelBundle(/*Insert=*/HasSpecConstArgs);
     if (KernelBundleImpPtr) {
       switch (KernelBundleImpPtr->get_bundle_state()) {
```

Comment:
Could you please also add a test to `sycl/unittests/SYCL2020`?

Reply:
This cannot currently be done due to a problem with PiImages not working without explicit an explicit kernel_bundle. Once that is fixed we can add this test, but until then this should be covered by https://github.com/intel/llvm-test-suite/pull/361.